receiver/kafkametricsreceiver/topic_scraper.go

```diff
@@ -0,0 +1,120 @@
+// Copyright  OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package kafkametricsreceiver
+
+import (
+	"context"
+	"regexp"
+	"time"
+
+	"github.com/Shopify/sarama"
+	"go.opentelemetry.io/collector/consumer/pdata"
+	"go.opentelemetry.io/collector/consumer/simple"
+	"go.opentelemetry.io/collector/receiver/scraperhelper"
+	"go.uber.org/zap"
+
+	"github.com/open-telemetry/opentelemetry-collector-contrib/receiver/kafkametricsreceiver/internal/metadata"
+)
+
+type topicScraper struct {
+	client       sarama.Client
+	logger       *zap.Logger
+	topicFilter  *regexp.Regexp
+	saramaConfig *sarama.Config
+	config       Config
+}
+
+func (s *topicScraper) Name() string {
+	return "topics"
+}
+
+func (s *topicScraper) shutdown(context.Context) error {
+	if !s.client.Closed() {
+		return s.client.Close()
+	}
+	return nil
+}
+
+func (s *topicScraper) scrape(context.Context) (pdata.ResourceMetricsSlice, error) {
+	topics, err := s.client.Topics()
+	metrics := simple.Metrics{
+		Metrics:                    pdata.NewMetrics(),
+		MetricFactoriesByName:      metadata.M.FactoriesByName(),
+		InstrumentationLibraryName: InstrumentationLibName,
+		Timestamp:                  time.Now(),
+	}
+	if err != nil {
+		return metrics.Metrics.ResourceMetrics(), err
+	}
+
+	var matchedTopics []string
+	for _, t := range topics {
+		if s.topicFilter.MatchString(t) {
+			matchedTopics = append(matchedTopics, t)
+		}
+	}
+	for _, topic := range matchedTopics {
+		partitions, err := s.client.Partitions(topic)
+		if err != nil {
+			continue
```

Comment:
Should it log?

Reply:
Would it be preferable to collect all these errors and log a summary at the end of the scrape operation before returning the successful metrics?
Like `error to occurred scraping {x} partitions in {y} topics`.
this would lose the error messages

Reply:
@jrcamp I've gone through some of the other receivers, and the closest thing I found to handling this is:
returning the successful metrics along with a [`consumererror.Combine`](https://github.com/open-telemetry/opentelemetry-collector/blob/3a9b4d8d4f5db6f442636b9d0bac803c72738f48/consumer/consumererror/combine.go#L29) of any errors that occur into a [`scrapererror.PartialScrapeError`](https://github.com/open-telemetry/opentelemetry-collector/blob/3a9b4d8d4f5db6f442636b9d0bac803c72738f48/receiver/scrapererror/partialscrapeerror.go#L21).

This will at least group all the errors together and not flood the output - it will still happen at every interval. 
But, these errors are not expected to occur often, any auth problems will be caught at initialization and connectivity errors will also happen before the loop is started.

Reply:
Yeah combining sounds good